cmd/duffle/relocate.go

```diff
@@ -0,0 +1,249 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pivotal/image-relocation/pkg/image"
+	"github.com/pivotal/image-relocation/pkg/pathmapping"
+	"github.com/pivotal/image-relocation/pkg/registry"
+
+	"github.com/deislabs/cnab-go/bundle"
+
+	"github.com/deislabs/duffle/pkg/duffle/home"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+const relocateDesc = `
+Relocates any docker and oci images referenced by a bundle, tags and pushes the images to a registry, and creates a new
+bundle with an updated image map.
+
+The --repository-prefix flag determines the repositories for the relocated images.
+Each image is tagged with a name starting with the given prefix and pushed to the repository.
+
+For example, if the repository-prefix is example.com/user, the image istio/proxyv2 is relocated
+to a name starting with example.com/user/ and pushed to a repository hosted by example.com.
+`
+
+type relocateCmd struct {
+	// args
+	inputBundle  string
+	outputBundle string
+
+	// flags
+	repoPrefix         string
+	inputBundleIsFile  bool
+	outputBundleIsFile bool
+
+	// context
+	home home.Home
+	out  io.Writer
+
+	// dependencies
+	mapping        pathmapping.PathMapping
+	registryClient registry.Client
+}
+
+func newRelocateCmd(w io.Writer) *cobra.Command {
+	relocate := &relocateCmd{out: w}
+
+	cmd := &cobra.Command{
+		Use:   "relocate [INPUT-BUNDLE] [OUTPUT-BUNDLE]",
+		Short: "relocate images in a CNAB bundle",
+		Long:  relocateDesc,
+		Example: `duffle relocate helloworld hellorelocated --repository-prefix example.com/user
+duffle relocate path/to/bundle.json relocatedbundle --repository-prefix example.com/user --input-bundle-is-file
+duffle relocate helloworld path/to/relocatedbundle.json --repository-prefix example.com/user --output-bundle-is-file`,
+		Args: cobra.ExactArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// validate --repository-prefix if it is set, otherwise allow flag omission to be diagnosed as such
+			if cmd.Flags().Changed("repository-prefix") {
+				if err := flagValidRepository("repository-prefix")(cmd); err != nil {
```

Comment:
Does this function really need to take this string argument and return a functor? As it is only used once, it can be simplified. You can also always check the repository prefix value set:
```go
PreRunE: func(cmd *cobra.Command, args []string) error {
if err := validateRepository(relocate. repoPrefix); err != nil{
    return err
}
```
with bellow, replacing `func flagValidRepository(flagName string) flagsValidator {`:
```go
func validateRepository(repository string) error {
    if strings.HasSuffix(repository, "/") || strings.Contains(repository, "//") {
        return fmt.Errorf(...)
...
```

Reply:
I got rid of the functor as suggested. (Functors made more sense in the project where this code originated since there was a lot of flag validation going on.)

Note that `PreRunE` deliberately falls through when the repository prefix is not specified so that we get the cobra error message for a missing required flag (for consistency). I've changed the comment to make this clearer.